### DISCUSSION TITLE
Support additional options in persist static method

### DISCUSSION THREAD

**Discussion by enzolupia** [jperasmus/stale-while-revalidate-cache#42] [Ideas]
Storage (i.e. redis) support additional configurations in set/write methods, i.e. ttl.

It would be useful to be able to pass those options to the `persist` static method.

Example:

Given the following configuration:
```
const swr = createStaleWhileRevalidateCache({
    maxTimeToLive:  60000,
    minTimeToStale:  20000,
    storage: {
      async getItem(cacheKey: string) {
        return redis.get(cacheKey);
      },
      async setItem(cacheKey: string, cacheValue: string | number | Buffer) {
        if (cacheValue) {
          await redis.set(cacheKey, cacheValue, 'PX',  60000);
        }
      },
      async removeItem(cacheKey: string) {
        await redis.del(cacheKey);
      },
    },
  });
```

It is not possible to override the ttl on `setItem` method on specific call to `persist`.

A solution is to extend the Storage interface, allowing at least an optional 'ttl' to be passed to the `setItem` method and of course extend the `persist` method as well to propagate the  new field to the setItem method.

What it should look like at the end is:

```
const swr = createStaleWhileRevalidateCache({
    maxTimeToLive:  60000,
    minTimeToStale:  20000,
    storage: {
      async getItem(cacheKey: string) {
        return redis.get(cacheKey);
      },
      async setItem(cacheKey: string, cacheValue: string | number | Buffer, ttl?: number) {
        if (cacheValue) {
          await redis.set(cacheKey, cacheValue, 'PX',  ttl ?? 60000);
        }
      },
      async removeItem(cacheKey: string) {
        await redis.del(cacheKey);
      },
    },
  });
```

``` 
swr.persist(key, value, 5000);
```

More than happy to open a PR for you to add at least a configurable ttl to the setItem static method

**Comment by jperasmus** [comment]
The "ttl" prop is too specific to the underlying storage implementation. Ie. Redis

The idea is that the storage config should be universal. Someone should be able to use a plain in-memory object as storage, or the browser's localStorage if they want. Both of those don't support this concept of ttl.

I can't remember exactly now, but you should be able to pass a complete "storage" config per request as override too. That way it can be specific to the the rest of the config for the request.

**Comment by enzolupia** [comment]
Yes, you are right, ttl is probably too specific.

Anyway, the storage config override can only be specified on the swr instance, not on the persist method specifically.

**Reply by jperasmus** [reply]
Okay, I will have to check that out. Unless you want to contribute a PR to allow the storage to be ootionally passed to the persist method?

**Reply by enzolupia** [reply]
Sure, I'll try to do it !

**Reply by enzolupia** [reply]
I have a solution implemented for you, instead of passing the whole storage (which probably doesn't make much sense) I added the possibility to pass additional generic options to the `persist` method.
I'm trying to push the code to a feature branch but it looks like I'm not allowed to do that.
```Permission to jperasmus/stale-while-revalidate-cache.git denied to enzolupia```

**Reply by jperasmus** [reply]
Hey, yes, you won't be able to push directly to this repo. You can fork this repo, and then create a PR from the fork to this one.

**Reply by enzolupia** [reply]
Great, I've just opened a pr for you, sorry I had to close my first pr and opened a second one.
